src/System.Private.CoreLib/src/System/Threading/ThreadPool.Unix.cs

```diff
@@ -2,23 +2,112 @@
+        /// in case if some threads get blocked while running user code.
+        /// </summary>
+        private static readonly int MaxThreadCount = 4 * ThreadPoolGlobals.processorCount;
+
+        /// <summary>
+        /// Semaphore that is used to release waiting thread pool workers when new work becomes available.
+        /// </summary>
+        private static SemaphoreSlim s_semaphore = new SemaphoreSlim(0);
+
+        /// <summary>
+        /// Number of worker threads created by the thread pool.
+        /// </summary>
+        private static volatile int s_workerCount = 0;
+
+        /// <summary>
+        /// This method is called to request a new thread pool worker to handle pending work.
+        /// </summary>
         internal static void QueueDispatch()
         {
-            // UNIXTODO: Threadpool
-            throw new NotImplementedException();
+            // For simplicity of the state management, we pre-create all thread pool workers on the first
+            // request and then use the semaphore to release threads as new requests come in.
+            if ((s_workerCount == 0) && Interlocked.Exchange(ref s_workerCount, MaxThreadCount) == 0)
+            {
+                for (int i = 0; i < MaxThreadCount; i++)
+                {
+                    if (!Interop.Sys.RuntimeThread_CreateThread(IntPtr.Zero /*use default stack size*/,
+                        AddrofIntrinsics.AddrOf<Interop.Sys.ThreadProc>(ThreadPoolDispatchCallback), IntPtr.Zero))
+                    {
+                        throw new OutOfMemoryException();
+                    }
+                }
+            }
+
+            // Release one thread to handle the new request
+            s_semaphore.Release(1);
         }
 
         internal static void QueueLongRunningWork(Action callback)
         {
-            // UNIXTODO: Threadpool
-            throw new NotImplementedException();
+            GCHandle gcHandle = GCHandle.Alloc(callback);
+
+            if (!Interop.Sys.RuntimeThread_CreateThread(IntPtr.Zero /*use default stack size*/,
+                AddrofIntrinsics.AddrOf<Interop.Sys.ThreadProc>(LongRunningWorkCallback), GCHandle.ToIntPtr(gcHandle)))
+            {
+                gcHandle.Free();
+                throw new OutOfMemoryException();
+            }
+        }
+
+        /// <summary>
+        /// This method is an entry point of a thread pool worker thread.
+        /// </summary>
+        [NativeCallable(CallingConvention = CallingConvention.StdCall)]
+        private static uint ThreadPoolDispatchCallback(IntPtr context)
+        {
+            RuntimeThread.InitializeThreadPoolThread();
+
+            do
+            {
+                // Handle pending requests
+                ThreadPoolWorkQueue.Dispatch();
+
+                try
+                {
+                    // Wait for new requests to arrive
+                    s_semaphore.Wait();
+                }
+                catch (ObjectDisposedException)
```

Comment:
I put this code in when I just started working on this change as a quick way to get rid of the “unreachable code detected” error. That was the only intention of this catch&break statement. I will remove it.